Sources/_SwiftSyntaxTestSupport/IncrementalParseTestUtils.swift

```diff
@@ -0,0 +1,127 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+import SwiftParser
+import XCTest
+
+/// This function is used to verify the correctness of incremental parsing
+/// containing three parts:
+/// 1. Round-trip on the incrementally parsed syntax tree.
+/// 2. verify that incrementally parsing the edited source base on the original source produces the same syntax tree as reparsing the post-edit file from scratch.
+/// 3. verify the reused nodes fall into expectations.
+public func verifyIncrementalParse(
+  originalSource: String,
+  editedSource: String,
+  concurrentEdits: ConcurrentEdits,
+  expectedReusedRanges: [(ByteSourceRange, SyntaxKind)]
+) {
+
+  let originalTree = Parser.parse(source: originalSource)
+
+  let reusedNodesCollector = IncrementalParseReusedNodeCollector()
+  let transition = IncrementalParseTransition(previousTree: originalTree, edits: concurrentEdits, reusedNodeDelegate: reusedNodesCollector)
+
+  let newTree = Parser.parse(source: editedSource)
+  let incrementallyParsedNewTree = Parser.parse(source: editedSource, parseTransition: transition)
+
+  // Round-trip
+  assertStringsEqualWithDiff(editedSource, "\(incrementallyParsedNewTree)")
+
+  // Substructure
+  let subtreeMatcher = SubtreeMatcher(Syntax(incrementallyParsedNewTree), markers: [:])
+  do {
+    try subtreeMatcher.assertSameStructure(Syntax(newTree))
+  } catch {
+    XCTFail("Matching for a subtree failed with error: \(error)")
+  }
```

Comment:
You could just make this function `throws` and then you don’t need to deal with `do {} catch {}` blocks

Reply:
Well, I just want to keep this the same as `assertParse`. We do error handling here so that we don't have to mark test cases throwable or add some do-catch blocks there, which looks nicer IMO.

Reply:
Well, I just want to keep this the same as `assertParse`. We do error handling here so that we don't have to mark test cases throwable or add some do-catch blocks there, which looks nicer IMO.

Reply:
Ah, yeah. Makes absolute sense. Thanks for clarifying.